Add content (README, licenses, TPN...) for the VMR (#14848)
dotnet/installer: src/SourceBuild/tarball/content/README.md

> # dotnet/dotnet - Home of the .NET VMR (experimental)

Reviewer: Have we decided (after the experiment concludes) if we're going to have one VMR per version? It could be useful to communicate our plans on this front.

Author: Fair point. We landed on "we're not going to name this `dotnet/dotnet8` and we'll see how things look later in the cycle to see if we need to split" so not sure we have a clear strategy to share?

Reviewer: I think a statement on "these are the two options we considered" could be useful. Versioning is a pretty relevant topic to any user.

Author: This is a good point. Poeple might just not know how new stuff is inside.

Should we maybe then declare that we expect that the VMR will host a full Linux build of .NET 8? So that people understand what version of .NET and when they can get out of it?

Should we explain that you can find the latest main in the main branch (and explain servicing branches)

Author: Also dealt with in the other thread

> - [Limitations](#limitations)

Reviewer: I think these TOCs unnecessarily block the reading flow. They are only useful if people click on them. I doubt people do.

- [Goals](#goals)
- [Building the VMR](#building-the-vmr)
- [List of components](#list-of-components)
- [Filing Issues](#filing-issues)
- [Useful Links](#useful-links)
- [.NET Foundation](#net-foundation)
- [License](#license)

> This repository is an **Experimental Virtual Monolithic Repository** that hosts all the code to help you build the entire .NET product end-to-end from source (often referred to as source-build).

Reviewer: ```suggestion
This repository is a **Virtual Monolithic Repository (VMR)**. It includes all the code needed to build the .NET SDK, mirrored from many atomic repos (like dotnet/runtime). It also includes [source-build](https://github.com/dotnet/source-build),  our whole-product build system. The VMR is currently experimental.
```

What this means:
- **Monolithic** - a join of multiple individual repositories that make up the whole product, such as [dotnet/runtime](https://github.com/dotnet/runtime) or [dotnet/sdk](https://github.com/dotnet/sdk).
- **Virtual** - not a replacement for the individual repositories that the product consists of but rather a mirror where sources of the individual repositories are synchronized into.

> - **Experimental** - not to be depended on as we reserve the right to delete the current instance and create a new, different one in its stead.

Reviewer 2: I think we should provide/link to a road map.  I think it is important to convey the commitment to this approach and when "dependencies" can be made on it.

Author: Do you mean this https://github.com/mmitche/arcade/blob/ub-docs/Documentation/UnifiedBuild/README.md?

Or do you mean a VMR-only more granular roadmap?

Reviewer 2: VMR granular.  In context of the description here, I think we should be including a timeframe in which we expect the "experimental" nature to be removed do that folks "can take a dependency" on this.

Reviewer: How about "by .NET 8 Preview 1". I hope it is before that, but that feels like a good outside bound to commit to now.

Author: I had a note to the `Limitations` section (plus a reference up here) with a link to the detailed roadmap that will be in the UB docs.

I tried to put a roadmap here as well but it was just getting very long and if I omitted parts, the rest didn't make a lot of sense so I think it's better to have the full thing in the UB?

Author: This has been resolved in the other thread. We now make it clear you can expect .NET 8 Linux builds from the VMR

In the VMR, you can find:
- [source files of each product repository](#list-of-components) which are mirrored inside of their respective directories under [`src/`](src/),
- tooling that enables [building the whole .NET product from source](https://github.com/dotnet/source-build) on Linux platforms,
- small customizations, in the form of patches, applied on top of the original code to make the build possible,
- *[in future]* E2E tests for the whole .NET product.

> More in-depth documentation about the VMR can be found in the [here (VMR Design And Operation)](https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild/VMR-Design-And-Operation.md#layout).  

Reviewer: ```suggestion
More in-depth documentation about the VMR can be found in [VMR Design And Operation](https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild/VMR-Design-And-Operation.md#layout).  
```

> See also the [dotnet/source-build](https://github.com/dotnet/source-build) repo for more information about the source-build itself.

Reviewer: ```suggestion
See also [dotnet/source-build](https://github.com/dotnet/source-build) for more information about our whole-product source-build.
```

## Limitations
**This is a work-in-progress and an experiment.**
There are considerable limitations to what is possible at the moment. For an extensive list of current limitations, please see [Temporary Mechanics](https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild/VMR-Design-And-Operation.md#temporary-mechanics).

### Supported platforms / builds

.NET Source-Build is supported on the oldest available .NET SDK feature update, and on Linux only.
For example, if both .NET 6.0.1XX and 6.0.2XX feature updates are available from [dotnet.microsoft.com](https://dotnet.microsoft.com/en-us/download/dotnet/6.0), Source-Build will only support 6.0.1XX.
For the latest information about Source-Build support for new .NET versions, please check our [GitHub Discussions page](https://github.com/dotnet/source-build/discussions) for announcements.

It is expected that other platforms will be supported in the future.

### Code flow
For the time being, the source code only flows one way - from the individual repos into the VMR.

> Changes done to the VMR are not automatically mirrored back. More details on this process:

Reviewer 2: I feel like we should provide clarity around the timeline of the code flow.  e.g. .NET 8.0 one way, .NET 9.0+ two way.  The linked documentation does not call this out either.

### Contribution

At this time, the VMR will not accept any changes outside of those necessary for enabling the supported builds to work.
Please, make the changes in the respective development repositories (e.g., [dotnet/runtime](https://github.com/dotnet/runtime) or [dotnet/sdk](https://github.com/dotnet/sdk)) and they will get synchronized into the VMR automatically.

## Goals

This repository eventually aims to become the place from which we release and service future versions of .NET.
The goal is to reduce complexity of the product construction process and thus enable partners and 3rd parties to easily build, test and modify .NET using their custom infrastructure.

Furthermore, we hope to solve other problems that the current multi-repo setup brings:
- Enable developers to make and test changes spanning multiple repositories.
- Fulfill requirements of .NET distro builders such as RedHat or Canonical to natively include .NET in their distribution repositories.

> - Make it possible for anyone to build the product without the need for currently used CI systems and inter-repo code flow orchestration. This should simplify scenarios such as client-run testing of bug fixes and improvements. The build should work in an offline environment too for certain platforms.

Author: Did you mean to remove the "Furthermore..." section?

> - Make it possible for anyone to build the product without the need for currently used CI systems and inter-repo code flow orchestration. This should simplify scenarios such as client-run testing of bug fixes and improvements. The build should work in an offline environment too for certain platforms.

Reviewer: Yes, the text you have in "Goals" looks good.

> Make it possible for anyone to build the product without the need for currently used CI systems and inter-repo code flow orchestration.

That's the text I was referring to.

> - Make it possible for anyone to build the product without the need for currently used CI systems and inter-repo code flow orchestration. This should simplify scenarios such as client-run testing of bug fixes and improvements. The build should work in an offline environment too for certain platforms.

Author: Gotcha, removed

- Enable the standard [down-/up-stream open-source model](https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild/VMR-Upstream-Downstream.md).
- More efficient pipeline for security fixes during the CVE pre-disclosure process.

## Building the VMR

Please note that **this repository is an experiment and a work-in-progress so it is possible that the build is broken**.

> For the latest information about Source-Build support, please check our [GitHub Discussions page](https://github.com/dotnet/source-build/discussions) for announcements.

Reviewer 2: ```suggestion
For the latest information about Source-Build support, please watch for announcements posted on our [GitHub Discussions page](https://github.com/dotnet/source-build/discussions).
```

### Prerequisites

The dependencies for building .NET from source can be found [here](https://github.com/dotnet/runtime/blob/main/docs/workflow/requirements/linux-requirements.md).

### Building

1. Clone the VMR.

   ```bash
   git clone https://github.com/dotnet/dotnet dotnet-dotnet
   ```

2. Prep the source to build on your distro. This downloads a .NET SDK and a number of .NET packages needed to build .NET from source.

    ```bash
    cd dotnet-dotnet
    ./prep.sh
    ```

3. Build the .NET SDK

    ```bash

>     ./build.sh --clean-while-building

Reviewer 2: Only online builds are supported at the moment.  This instruction should include the `--online` option.

Author: Should I call that out in the `Limitations`?

More in-depth documentation about the VMR can be found in [VMR Design And Operation](https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild/VMR-Design-And-Operation.md#layout).  
See also [dotnet/source-build](https://github.com/dotnet/source-build) for more information about our whole-product source-build.

## Limitations

**This is a work-in-progress and an experiment.**
There are considerable limitations to what is possible at the moment. For an extensive list of current limitations, please see [Temporary Mechanics](https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild/VMR-Design-And-Operation.md#temporary-mechanics).

The VMR is expected to become non-experimental by .NET 8 Preview 1 (Februrary, 2023) 
This means it won't be short-lived anymore and we won't be reserving the right to delete and re-create it anymore.
Other limitations might apply until the .NET 9 timeframe.
See the [Unified Build roadmap](https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild) for more details.

### Supported platforms / builds

.NET Source-Build is supported on the oldest available .NET SDK feature update for each major release, and on Linux only.

> For example, if .NET `6.0.1xx`, `6.0.2xx`, and `7.0.1xx` feature updates are available from [dotnet.microsoft.com](https://dotnet.microsoft.com/en-us/download/dotnet/6.0), Source-Build will support `6.0.1xx` and `7.0.1xx`.

Author: ```suggestion
For example, if .NET `6.0.1xx`, `6.0.2xx`, and `7.0.1xx` feature updates are available from [dotnet.microsoft.com](https://dotnet.microsoft.com/en-us/download/dotnet/6.0), Source-Build will support `6.0.1xx` and `7.0.1xx`.

That being said, the VMR has only been created after .NET 7.0 had been released so we only plan to use it for supporting .NET 8.0 and further. This means that the VMR's main follows the main of the product repositories (see [Source Synchronization Process](https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild/VMR-Design-And-Operation.md#source-synchronization-process)). The VMR will also follow suite with the release branches for .NET 8 (e.g. `release/8.0.1xx-preview1` and similar).
```

Author: @MichaelSimons can you help me formulating this paragraph? This is related to the other open threads in this PR and I am wondering whether this doesn't deserve its own section describing what people can actually find in the VMR

Author: We have not explicitly talked about this but I always somehow thought about it that way. Only seems reasonable?

Author: I have rewritten a bit. Can you have a look?

Reviewer 2: > Only seems reasonable to synchronize all release branches of installer into their respective ones in the VMR? How else would we do servicing?

Agreed - I wanted to make sure everyone is thinking the same.

> I have rewritten a bit. Can you have a look?

Done - Added comments inline.

It is expected that other platforms will be supported in the future.

### Online build only

Building the product offline is not fully working at the moment. The `--online` switch is needed when [building](#building) as not all dependencies are currently built from source.

### Code flow
For the time being, the source code only flows one way - from the individual repos into the VMR.
Changes done to the VMR are not automatically mirrored back. More details on this process:

- [Source Synchronization Process](https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild/VMR-Design-And-Operation.md#source-synchronization-process)
- [Moving Code and Dependencies between the VMR and Development Repos](https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild/VMR-Design-And-Operation.md#moving-code-and-dependencies-between-the-vmr-and-development-repos)

We expect the code flow to start working both ways in the .NET 9 timeframe.
See the [Unified Build roadmap](https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild) for more details.

### Contribution

> At this time, the VMR will not accept any changes outside of those necessary for enabling the supported builds to work.

Author: Yeah, with mirroring the content folder, this was outdated.

Please, make the changes in the respective development repositories (e.g., [dotnet/runtime](https://github.com/dotnet/runtime) or [dotnet/sdk](https://github.com/dotnet/sdk)) and they will get synchronized into the VMR automatically.

> ## Goals

Reviewer 2: What do you think about moving the `goals` above `limitations`?  I feel like starting out with limitations so early is a detractor or what this repo is trying to accomplish (e.g. its goals).

Author: Yup, makes sense

The main purpose of this repository is to have all source code necessary to build the whole .NET product from source contained in any given commit.

Further, the VMR eventually aims to become the place from which we release and service future versions of .NET.
This is to reduce the complexity of the product construction process and thus enable partners and 3rd parties to easily build, test and modify .NET using their custom infrastructure.

Furthermore, we hope to solve other problems that the current multi-repo setup brings:

> - Enable developers to make and test changes spanning multiple repositories.

Reviewer 2: Are these in priority order?  They don't align with what I think the priority should be.  I see this one as being towards the bottom.

Author: They are not numbered but I guess the order matters the same. I re-shuffled them.

### Prerequisites

The dependencies for building .NET from source can be found [here](https://github.com/dotnet/runtime/blob/main/docs/workflow/requirements/linux-requirements.md).

### Building

1. **Clone the VMR**

   ```bash
   git clone https://github.com/dotnet/dotnet dotnet-dotnet
   ```

2. **Prep the source to build on your distro**  
   This downloads a .NET SDK and a number of .NET packages needed to build .NET from source.

    ```bash
    cd dotnet-dotnet
    ./prep.sh
    ```

3. **Build the .NET SDK**

    ```bash
    ./build.sh --clean-while-building --online
    ```

    This builds the entire .NET SDK from source.
    The resulting SDK is placed at `artifacts/x64/Release/dotnet-sdk-7.0.100-your-RID.tar.gz`.

    Currently, the `--online` flag is required to allow NuGet restore from online sources during the build.
    This is useful for testing unsupported releases that don't yet build without downloading pre-built binaries from the internet.

    Run `./build.sh --help` to see more information about supported build options.

4. *(Optional)* **Unpack and install the .NET SDK**

    ```bash
    mkdir -p $HOME/dotnet

>     tar zxf artifacts/x64/Release/dotnet-sdk-7.0.100-your-RID.tar.gz -C $HOME/dotnet

Author: ```suggestion
    tar zxf artifacts/x64/Release/dotnet-sdk-8.0.100-your-RID.tar.gz -C $HOME/dotnet
```

    ln -s $HOME/dotnet/dotnet /usr/bin/dotnet
    ```
    
    To test your source-built SDK, run the following:

    ```bash
    dotnet --info
    ```

## List of components

To enable full offline source-building of the VMR, we have no other choice than to synchronize all the necessary code into the VMR. This also includes any code referenced via git submodules. More details on why we need to do and how we do that can be found here:
- [Strategy for managing external source dependencies](https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild/VMR-Strategy-For-External-Source.md)
- [Source Synchronization Process](https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild/VMR-Design-And-Operation.md#source-synchronization-process)

### Detailed list

<!-- component list beginning -->

> Auto-generated list of components will go here

<!-- component list end -->

The repository also contains a [JSON manifest](https://github.com/dotnet/dotnet/blob/main/src/source-manifest.json) listing all components in a machine-readable format.

## Filing Issues

> This repo should only contain issues that are tied to the VMR construction/build itself. For other issues, please file them to their appropriate development repos.

Reviewer 2: > This repo should only contain issues that are tied to the VMR construction/build itself.

I thought we were going to hold off on enabling issues initially?

Author: Changed to:

> This repo does not accept issues as of now. Please file issues to the appropriate development repos.
> For issues with the VMR itself, please use the [source-build repository](https://github.com/dotnet/source-build).

## Useful Links

- Design documentation for the VMR - a set of documents describing the high-level design and the why's and how's
  - [Design and Operation](https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild/VMR-Design-And-Operation.md)
  - [Upstream/Downstream Relationships](https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild/VMR-Upstream-Downstream.md)
  - [Code and Build Workflow](https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild/VMR-Code-And-Build-Workflow.md)
  - [Strategy for Managing External Source Dependencies](https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild/VMR-Strategy-For-External-Source.md)
- [.NET Source-Build](https://github.com/dotnet/source-build)
- [What is .NET](https://dotnet.microsoft.com)

## .NET Foundation

.NET Runtime is a [.NET Foundation](https://www.dotnetfoundation.org/projects) project.

## License

> .NET (including the runtime repo) is licensed under the [MIT](LICENSE.TXT) license.

Author: ```suggestion
.NET is licensed under the [MIT](LICENSE.TXT) license.
```

- Simplify scenarios such as client-run testing of bug fixes and improvements. The build should work in an offline environment too for certain platforms.
- Enable developers to make and test changes spanning multiple repositories.
- More efficient pipeline for security fixes during the CVE pre-disclosure process.

## Limitations

**This is a work-in-progress and an experiment.**
There are considerable limitations to what is possible at the moment. For an extensive list of current limitations, please see [Temporary Mechanics](https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild/VMR-Design-And-Operation.md#temporary-mechanics).

The VMR is expected to become non-experimental by .NET 8 Preview 1 (Februrary, 2023) 
This means it won't be short-lived anymore and we won't be reserving the right to delete and re-create it anymore.
Other limitations might apply until the .NET 9 timeframe.
See the [Unified Build roadmap](https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild) for more details.

### Supported platforms / builds

> .NET Source-Build is supported on the oldest available .NET SDK feature update for each major release, and on Linux only.

Reviewer 2: We should drop the oldest available .NET SDK part all together.  The only part that is relevant is that .NET source build only supported Linux.

Author: Removed. Additionally, I moved the "we will have main and release branches" above under the "in the VMR you can find" section.

[`a0aaa40` (#14848)](https://github.com/dotnet/installer/pull/14848/commits/a0aaa409b850090520c9c889e3c8c1da4a4ae9b5)

For example, if .NET `6.0.1xx`, `6.0.2xx`, `7.0.1xx`, and `7.0.2xx` feature updates are available from [dotnet.microsoft.com](https://dotnet.microsoft.com/en-us/download/dotnet/6.0), Source-Build will support `6.0.1xx` and `7.0.1xx`.

> That being said, the VMR has only been created after .NET 7.0 had been released so it will contain .NET 8.0 and higher.

Reviewer 2: Perhaps the previous paragraph can be combined here.

```suggestion
The VMR only support .NET 8.0 and higher.  Additionally source-build currently support Linux only.
```

    ```bash
    mkdir -p $HOME/dotnet
    tar zxf artifacts/x64/Release/dotnet-sdk-8.0.100-your-RID.tar.gz -C $HOME/dotnet
    ln -s $HOME/dotnet/dotnet /usr/bin/dotnet
    ```
    
    To test your source-built SDK, run the following:

    ```bash
    dotnet --info
    ```

## List of components

To enable full offline source-building of the VMR, we have no other choice than to synchronize all the necessary code into the VMR. This also includes any code referenced via git submodules. More details on why we need to do and how we do that can be found here:
- [Strategy for managing external source dependencies](https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild/VMR-Strategy-For-External-Source.md)
- [Source Synchronization Process](https://github.com/dotnet/arcade/blob/main/Documentation/UnifiedBuild/VMR-Design-And-Operation.md#source-synchronization-process)

### Detailed list

<!-- component list beginning -->

> Auto-generated list of components will go here

<!-- component list end -->

The repository also contains a [JSON manifest](https://github.com/dotnet/dotnet/blob/main/src/source-manifest.json) listing all components in a machine-readable format.

> ## Filing Issues

Author: TODO: Based on where the canonical issue/discussion will be, we should guide people there (e.g. source-build repo discussions)
